quantile-regression/README.md

```diff
@@ -0,0 +1,69 @@
+# Quantile Regression
+
+This discusses how we can include quantile regression in tidymodels. 
+
+The discussion below is organized by package
+
+## parsnip
+
+Users will probably have to specify the quantiles at the specification time (see section on new models). 
+
+### Engines and Modes
+
+The main question is: Should we make new engines available via `set_engine()` or create a new mode of `"quantile regression"`? 
+
+Pros of `set_engine()`: 
+
+* Most (but not all) quantile regression predictions will be for numeric modes. 
+* There is not much difference between quantile models that require a new mode (unlike censored regression). For example, our yardstick functions to compute loss for quantile regression can ingest the list column of predictions (similar to dynamic survival models). 
+* Where do we specify the list of quantiles? It would be suboptimal to add a main option to each function. The users could set it in `set_engine(),`, but that would require more specialized parsing of the return value of that function. 
+
+Cons of `set_engine()`: 
+
+* There could be confusion around the type of model and prediction being made. Suppose that `rpart` could make quantile predictions (it cannot). Would use an engine of “part” result in the regular CART model or one optimized via quantile loss? 
+* Since the list of quintiles would be specified upfront, it would be advantageous to do this with the new mode (`set_mode(“quantgreg,” quantiles = (1:9) / 10)`). 
+
+To test using a new mode, there is a [parsnip branch](https://github.com/tidymodels/parsnip/tree/quantile-mode) that enables that so you can use: 
+
+```
+pak::pak(c("tidymodels/parsnip@quantile-mode"), ask = FALSE)
+```
+
+### New Models
+
+Some packages already fit these models, notably quantreg and quantregForest. 
+
+We can also make our own (I might do this with brulee). If so, the model functions should try to emulate the main arguments that we have. So for a neural network, use `hidden_units`, `penalty`, etc. 
+
+Many models need a specific training run for each quantile value (since the loss depends on it). For this reason, we would have users specify the required quantiles when they specify the model. 
+
+We should make a general class for these models in an engine-specific package. 
+
+#### Predictions
+
+We currently note tin `?predict.model_fit`: 
+
+> For `type = "quantile"`, the tibble has a `.pred` column, which is a list-column. Each list element contains a tibble with columns `.pred` and `.quantile` (and perhaps other columns).
+
+This should be changed so the list column is called `.pred_quantile` if we want `.pred` to contain the 50% quantile results. 
+
+The format of the list column would be 
+
+```
+tibble(.quantile = numeric(), .pred_quantile = numeric())
+```
+
+## yardstick
+
+We need one or more model performance metrics for quantile loss that work across all quantiles. They will need a new class too (say `“quantile_metric”`). 
```

Comment:
I think the most obvious metric is just to add up the quantile loss over all quantile levels. This is given in (for example) equation (10) of these notes: https://www.stat.berkeley.edu/~ryantibs/statlearn-s23/lectures/calibration.pdf.

As the notes explain, this is also equivalent (up to a factor of 2) to weighted interval score (WIS). That is defined in equation (8) and the equivalence is stated in equation (11).

Some people in the field do not know of this equivalence and I've even seen papers reporting both quantile loss and WIS as separate columns of a predictive perrformance metrics table. So it'd be good for this package to clearly state they are equivalent! Perhaps in the documentation.

WIS is nice because it offers a complementary view on the same equivalent metric. As explained in this paper: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008618, you can decompose the score by inspecting the two summands separately (summed up over all levels), and interpret these loosely as a spread/sharpness measure (first one) and calibration measure (second one). Some people like looking at that as consituents of error.

Other than quantile loss and equivalently WIS, I'm not aware of any other score based on predicted quantiles that is proper. People in forecast scoring care a lot about propriety. However, I'll just note that it would be reasonable in general to additionally look at absolute error (AE) using the predicted 50% quantile (predicted median) if it's available in the list of quantiles, and also to look at coverage (empirical coverage of each quantile versus the nominal coverage). I think people often do this in practice.

Finally, in an off-the-cuff remark, I'll note that I'm becoming more and more bothered by quantile loss/WIS as a general go-to metric. I think it can be too lenient for underdispersed forecasters. I have an ongoing project with students exploring this. I don't have anything concrete to recommend as an alternative at the moment, maybe we will at some future point.